File: 1059-all-paths-from-source-lead-to-destination/1059-all-paths-from-source-lead-to-destination.cpp
class Solution {
public:
    int visited[100001]; 
    bool dfs(vector<vector<int>>& adj, int source, int destination){
        if(visited[source] == 1) return false;  //found a cycle 
        if(visited[source] == 2) return true; 
        visited[source] = 1; 
        for(int next : adj[source]){
            if(!dfs(adj,next,destination)) return false; 
        }
        visited[source] = 2; 
        if(adj[source].size() == 0 && source != destination) return false; 
        return true; 
    }
    bool leadsToDestination(int n, vector<vector<int>>& edges, int source, int destination) {
        vector<vector<int>> adj(n); 
        for(vector<int>& v : edges){
            int from = v[0], to = v[1]; 
            adj[from].push_back(to); 
        }

        return dfs(adj,source,destination); 
    }
};

File: 1059-all-paths-from-source-lead-to-destination/1059-all-paths-from-source-lead-to-destination.java
class Solution {
    int[] visited = new int[100001]; 
    public boolean dfs(List<List<Integer>> adj, int source, int destination){
        if(visited[source] == 1) return false; //made cycle 
        if(visited[source] == 2) return true; 
        visited[source] = 1;  
        for(int next : adj.get(source)){
            if(!dfs(adj,next,destination)){
                return false; 
            }
        }
        visited[source] = 2; 
        if(adj.get(source).size() == 0 && source != destination) return false; 
        return true; 
    }
    public boolean leadsToDestination(int n, int[][] edges, int source, int destination) {
        List<List<Integer>> adj = new ArrayList<>();
        for(int i = 0; i < n; i++){
            adj.add(new ArrayList<>()); 
        }

        for(int[] e : edges){
            int from = e[0]; 
            int to = e[1]; 
            adj.get(from).add(to); 
        }


        return dfs(adj,source,destination); 
    }
}